### ISSUE TITLE
[clang][analyzer] False-positive null dereference through a const function pointer

### ISSUE THREAD

**Issue by bozicrHT** [llvm/llvm-project#216983]
**Description:**
The analyzer loses caller constraints when a function is invoked through a const function pointer:
```c
static void (*const callback)(unsigned) = notify;
```
Reproducer can be found here: <https://compiler-explorer.com/z/z6vrsKM37>. `backend_notify()` restricts `id` to `[1, 4]` before calling `callback(id)`. However, the analyzer does not resolve `callback` to `notify`, analyzes `notify()` separately with an unconstrained argument, and reports a false-positive null dereference. Calling `notify(id)` directly produces no warning.

Additionally `clang_analyzer_eval(callback == notify);` produces `UNKNOWN`. Since `callback` is a const pointer I would expect this to produce `TRUE`.

**Possible solution:**
I think that the function `SValBuilder::getConstantVal(const Expr *E)` should include logic to recognize direct function-address initializers such as `notify` and `&notify`, and represent them using the existing `FunctionCodeRegion` SVal via `getFunctionPointer()`.

**Comment by Snape3058** [comment]
I remembered that we had some discussions on whether to model global variables before. The execution output shows the negative answer. IMO, modeling the initial assignment of global const variables seems reasonable.

**Comment by NagyDonat** [comment]
> IMO, modeling the initial assignment of global const variables seems reasonable.

If I recall correctly we already have logic that can interpret the initializer of a global constant variable to resolve its value. To fix this bug, it is probably sufficient to tweak that logic to support functions and function pointers (which would probably require some simple special case logic to work around casts etc.).
